### ISSUE TITLE
Which Frontend Framework?

### ISSUE THREAD

**Issue by grantcarthew** [grantcarthew/rjq-ui-design#1]
So @sagivf asked why not Angular?

As I have stated, I don't want to build this solution myself, so with that in mind my opinion doesn't really count.

My opinion is with Angular you have to learn a specific API rather than how to use it with normal js. Vue.js seems to be more js focused. Same with Aurelia.

I have done a project with both Angular, Aurelia, and Vue.js and much prefer Vue.js.

Over to you guys.

**Comment by sagivf** [comment]
I am with you on the specific API part, I have long been wondering if going for a simpler framework or just Vanilla JS (though they do say "if you don't use a framework you'll end up building one").

Question is does Vue do that or is it another set of of api's to learn?
Since I've only done Angular in the last years I can't really say and I'll take your word on it, lets give
it a shot 👍

**Comment by TomKaltz** [comment]
I abandoned Angular many years ago when I had to do $watch everywhere and
haven't turned back.  Vue has a learning curve but it's a beautiful
framework once you get to know it.  I am swamped all week or I would put
together the boilerplate for this.  If you guys can wait until next week I
can get the boilderplate for this going and a repo up.  Should the
front-end and backend be separate repos?  My gut tells me no.  Any opinions?

Thomas R. Kaltz III

(586) 214-7150 cell
(866) 855-6350 fax

**Comment by sagivf** [comment]
I don't think there is that much of a rush so if you can get the boll rolling next week that would be great by me. If you don't seem to have the time let me/us now.

I would agree that nowadays angular1 has some clear drawbacks compared to other frameworks. I don't think that's the case for angular2.

You both are now angular and vue and prefer vue so lets for it, maybe i'll fall in love 😄 

I think one repo would be a good start, we can always split it up later.

**Comment by grantcarthew** [comment]
OK, lets make a decision on `Vue.js`.

Please see #2 for location discussion.

What about style?

I would go for [Foundation](http://foundation.zurb.com/sites.html). I like it better than [Bootstrap](http://getbootstrap.com/) however that's up to you guys to choose.

**Comment by sagivf** [comment]
Ive never used Foundation but happy to try it out.

**Comment by sagivf** [comment]
@TomKaltz @chrisabrams @grantcarthew Ive been a little busy lately but next week I start to free up. I want to plan my week and would like to dedicate a day to this project. Have any of you started anything? If not I can start working on it. I will probably start with a mocked front end.

**Comment by grantcarthew** [comment]
@sagivf Ok, sounds good.
If you are going to go with Vuejs and Foundation I have been working on this for the last two weeks in my personal project.

Here are some points of interest:

*   Use the [vue-cli](https://vuejs.org/v2/guide/installation.html#CLI) - `npm install vue-cli` with the [webpack template](https://github.com/vuejs-templates/webpack).
*   Use standard `npm install -D standard`
*   For Foundation-Sites: `npm install -D foundation-sites what-input motion-ui`
*   Font-Awesome? `npm install -D font-awesome`
*   Sass support? `npm install node-sass sass-loader`
*   For configuring Foundation: https://github.com/hal0gen/vue-foundation

I got some good points from that last bulletpoint however here are some changes I made:

In `build/utils.js` to add scss variables and references:

```js

  let sassOptions = {
    indentedSyntax: true
  }
  let scssOptions = {
    includePaths: [
      '~src/styles'
    ],
    data: '@import "~src/styles/app";'
  }

  // http://vuejs.github.io/vue-loader/en/configurations/extract-css.html
  return {
    css: generateLoaders(),
    postcss: generateLoaders(),
    less: generateLoaders('less'),
    sass: generateLoaders('sass', sassOptions),
    //  Make custom SASS available to all components https://github.com/webpack-contrib/sass-loader
    scss: generateLoaders('sass', scssOptions),
    stylus: generateLoaders('stylus'),
    styl: generateLoaders('stylus')
  }


```

Here is what the `src/main.js` looks like:

```js

// The Vue build version to load with the `import` command
// (runtime-only or standalone) has been set in webpack.base.conf with an alias.
import Vue from 'vue'
import App from './App'
import router from './router'
import jQuery from 'jquery'
window.jQuery = window.$ = jQuery

require('motion-ui')
require('what-input')
require('foundation-sites')

/* eslint-disable no-new */
new Vue({
  el: '#app',
  router,
  template: '<App/>',
  components: { App }
})


```

And `styles/app.scss`:

```js

// This file gets imported in every component by the scss cssLoader
// It is referenced by the sass compiler within build/utils.js
// https://github.com/vuejs/vue-loader/issues/328

// Load Foundation
@import './settings';
@import '~foundation-sites/scss/foundation';

// Load scss variables and settings
// These are your own custom variables in scss files under styles.
@import './layout';
@import './palette';
@import './colours';


```

I don't know if you need this `mixins/foundation.js` yet, still working on things:

```js

// Foundation mixin

module.exports = {
  mounted () {
    // eslint-disable-next-line
    $(this.$el).foundation();
  },
  unmounted () {
    // eslint-disable-next-line
    $(this.$el).foundation.destroy();
  }
}


```

Anyway, that will get you started if you want to go with Vuejs and Foundation.

My email is firstname@lastname.net replacing firstname with my first name etc. If you want more detail outside of github.

**Comment by grantcarthew** [comment]
Oh, another thing. If you want to start a repo under your name do so. You could also just add code here for now in the design repo. That's up to you.

If you are starting a new repo you will need a name. I tend to go with functional names like `rethinkdb-job-queue` however adding `ui` on the end of that is making it even longer. You could use a shorter name like `rjq-ui` or make up a completely new name like `donuts`  or something silly like that.

The name is not overly important because once it is working I will link to it on the `rethinkdb-job-queue` readme page.

**Comment by grantcarthew** [comment]
Update to my above post with example. The owner of the `vue-foundation` repo bugged me for a PR so I have fixed a lot of the issues in his template. Check the pull requests on his repo if it is not merged by the time you get to it.

https://github.com/hal0gen/vue-foundation

**Comment by sagivf** [comment]
Hey guys,
After some discussions with @grantcarthew i've started a new project - https://github.com/sagivf/rjq-ui.

Currently it's just a few views starting/prototyping what I think this project could look like.

You can see a mocked version in https://sagivf.github.io/rjq-ui/public/#/
It's using a project Ive worked on with friends here and there https://github.com/outbrain/Leonardo.
`Leonardo` is still in flux and there is a improved less buggy version on the way. If you need help using it let me know.

A few points:
* I would like everything mocked via Leonardo before we start working on a backend. I think it's a good way to define the needs in the UI and then implement.
* Feel free to open issues and add ideas, we can still spec things out in this project if you like.
* Still working on integrating Leonardo with webpack properly.

**Comment by grantcarthew** [comment]
Looks great Sagiv

**Comment by sagivf** [comment]
Also if anyone is proficient in `vue` or `foundation` or both.
I could use some help on this: https://github.com/sagivf/rjq-ui/issues/1

**Comment by chrisabrams** [comment]
The leonardo tools is very cool...ive really needed something simple like this!

**Comment by sagivf** [comment]
@chrisabrams Glad you like it :) give us a ⭐️ 
There is a much nicer (and less buggy) version on the way, it was written not to depend on angular and coming soon.
